`NullPointerException` in `fzC_universe.<clinit>` for empty array

This
```
x is
b array x := []
```
produces
```
 > ./build/bin/fz tj0.fz

error 1: java.lang.ExceptionInInitializerError
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at dev.flang.be.jvm.Runner.runMain(Runner.java:120)
	at dev.flang.be.jvm.JVM$CompilePhase$4.finish(JVM.java:510)
	at dev.flang.be.jvm.JVM.lambda$createCode$0(JVM.java:829)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
	at java.base/java.util.stream.ReferencePipeline$Head.forEachOrdered(ReferencePipeline.java:772)
	at dev.flang.be.jvm.JVM.createCode(JVM.java:820)
	at dev.flang.be.jvm.JVM.compile(JVM.java:805)
	at dev.flang.tools.Fuzion$Backend$3.process(Fuzion.java:163)
	at dev.flang.tools.Fuzion$Backend.processFrontEnd(Fuzion.java:444)
	at dev.flang.tools.Fuzion.lambda$parseArgsForBackend$3(Fuzion.java:908)
	at dev.flang.tools.Tool.lambda$run$0(Tool.java:145)
	at dev.flang.util.Errors.runAndExit(Errors.java:845)
	at dev.flang.tools.Tool.run(Tool.java:145)
	at dev.flang.tools.Fuzion.main(Fuzion.java:557)
Caused by: java.lang.NullPointerException: Cannot read the array length because "null" is null
	at fzC_universe.<clinit>(Unknown Source)
	... 18 more


*** fatal errors encountered, stopping.

```

--- a/src/dev/flang/be/jvm/CodeGen.java
+++ b/src/dev/flang/be/jvm/CodeGen.java
@@ -998,9 +998,7 @@ class CodeGen
                     .andThen(c._v0)                                // T[], T[], idx, const-data-code
                     .andThen(jt.xastore());                        // T[]
                 }
-              yield jt == PrimitiveType.type_void
-                ? new Pair<>(result, Expr.UNIT)
-                : _jvm.const_array(constCl, result);
+              yield _jvm.const_array(constCl, result, elCount);
             }
           else if (!_fuir.clazzIsChoice(constCl))
             {

--- a/src/dev/flang/be/jvm/CodeGen.java
+++ b/src/dev/flang/be/jvm/CodeGen.java
@@ -998,7 +998,9 @@ class CodeGen
                     .andThen(c._v0)                                // T[], T[], idx, const-data-code
                     .andThen(jt.xastore());                        // T[]
                 }
-              yield _jvm.const_array(constCl, result);
+              yield jt == PrimitiveType.type_void
+                ? new Pair<>(result, Expr.UNIT)
+                : _jvm.const_array(constCl, result);
             }
           else if (!_fuir.clazzIsChoice(constCl))
             {

--- a/src/dev/flang/be/jvm/JVM.java
+++ b/src/dev/flang/be/jvm/JVM.java
@@ -1482,7 +1482,7 @@ should be avoided as much as possible.
                        .andThen(Expr.invokeStatic(Names.RUNTIME_CLASS,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_8,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_8_SIG,
-                                                  PrimitiveType.type_byte.array())));
+                                                  PrimitiveType.type_byte.array())), bytes.length);
   }
 
 
@@ -1499,7 +1499,7 @@ should be avoided as much as possible.
                        .andThen(Expr.invokeStatic(Names.RUNTIME_CLASS,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_I16,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_I16_SIG,
-                                                  PrimitiveType.type_byte.array())));
+                                                  PrimitiveType.type_byte.array())), bytes.length / 2);
   }
 
 
@@ -1516,7 +1516,7 @@ should be avoided as much as possible.
                        .andThen(Expr.invokeStatic(Names.RUNTIME_CLASS,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_U16,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_U16_SIG,
-                                                  PrimitiveType.type_byte.array())));
+                                                  PrimitiveType.type_byte.array())), bytes.length / 2);
   }
 
 
@@ -1533,7 +1533,7 @@ should be avoided as much as possible.
                        .andThen(Expr.invokeStatic(Names.RUNTIME_CLASS,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_32,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_32_SIG,
-                                                  PrimitiveType.type_byte.array())));
+                                                  PrimitiveType.type_byte.array())), bytes.length / 4);
   }
 
 
@@ -1550,7 +1550,7 @@ should be avoided as much as possible.
                        .andThen(Expr.invokeStatic(Names.RUNTIME_CLASS,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_64,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_64_SIG,
-                                                  PrimitiveType.type_byte.array())));
+                                                  PrimitiveType.type_byte.array())), bytes.length / 8);
   }
 
 
@@ -1567,7 +1567,7 @@ should be avoided as much as possible.
                        .andThen(Expr.invokeStatic(Names.RUNTIME_CLASS,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_F32,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_F32_SIG,
-                                                  PrimitiveType.type_byte.array())));
+                                                  PrimitiveType.type_byte.array())), bytes.length / 4);
   }
 
 
@@ -1584,7 +1584,7 @@ should be avoided as much as possible.
                        .andThen(Expr.invokeStatic(Names.RUNTIME_CLASS,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_F64,
                                                   Names.RUNTIME_INTERNAL_ARRAY_FOR_ARRAY_F64_SIG,
-                                                  PrimitiveType.type_byte.array())));
+                                                  PrimitiveType.type_byte.array())), bytes.length / 8);
   }
 
 
@@ -1595,7 +1595,7 @@ should be avoided as much as possible.
    *
    * @param arr expr producing the java array, e.g. double[].
    */
-  Pair<Expr, Expr> const_array(int arrayCl, Expr arr)
+  Pair<Expr, Expr> const_array(int arrayCl, Expr arr, int len)
   {
     var internalArray  = _fuir.lookup_array_internal_array(arrayCl);
     var fuzionSysArray = _fuir.clazzResultClazz(internalArray);
@@ -1606,11 +1606,9 @@ should be avoided as much as possible.
       .andThen(new0(fuzionSysArray))                  //        cs, cs, fsa
       .andThen(Expr.DUP)                              //        cs, cs, fsa, fsa
       .andThen(arr)                                   //        cs, cs, fsa, fsa, arr
-      .andThen(Expr.DUP_X2)                           //        cs, cs, arr, fsa, fsa, arr
-      .andThen(putfield(data))                        //        cs, cs, arr, fsa
-      .andThen(Expr.DUP_X1)                           //        cs, cs, fsa, arr, fsa
-      .andThen(Expr.SWAP)                             //        cs, cs, fsa, fsa, arr
-      .andThen(Expr.ARRAYLENGTH)                      //        cs, cs, fsa, fsa, len
+      .andThen(putfield(data))                        //        cs, cs, fsa
+      .andThen(Expr.DUP)                              //        cs, cs, fsa, fsa
+      .andThen(Expr.iconst(len))                      //        cs, cs, fsa, fsa, len
       .andThen(putfield(length))                      //        cs, cs, fsa
       .andThen(putfield(internalArray))               //        cs
       .is(_types.javaType(arrayCl));                  //        -

--- a/tests/reg_issue2478/Makefile
+++ b/tests/reg_issue2478/Makefile
@@ -1,0 +1,25 @@
+# This file is part of the Fuzion language implementation.
+#
+# The Fuzion language implementation is free software: you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as published
+# by the Free Software Foundation, version 3 of the License.
+#
+# The Fuzion language implementation is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
+# License for more details.
+#
+# You should have received a copy of the GNU General Public License along with The
+# Fuzion language implementation.  If not, see <https://www.gnu.org/licenses/>.
+
+
+# -----------------------------------------------------------------------
+#
+#  Tokiwa Software GmbH, Germany
+#
+#  Source code of Fuzion test Makefile
+#
+# -----------------------------------------------------------------------
+
+override NAME = reg_issue2478.fz
+include ../positive.mk

--- a/tests/reg_issue2478/reg_issue2478.fz
+++ b/tests/reg_issue2478/reg_issue2478.fz
@@ -1,0 +1,25 @@
+# This file is part of the Fuzion language implementation.
+#
+# The Fuzion language implementation is free software: you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as published
+# by the Free Software Foundation, version 3 of the License.
+#
+# The Fuzion language implementation is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
+# License for more details.
+#
+# You should have received a copy of the GNU General Public License along with The
+# Fuzion language implementation.  If not, see <https://www.gnu.org/licenses/>.
+
+
+# -----------------------------------------------------------------------
+#
+#  Tokiwa Software GmbH, Germany
+#
+#  Source code of Fuzion test
+#
+# -----------------------------------------------------------------------
+
+x is
+b array x := []

--- a/tests/reg_issue2478/reg_issue2478.fz
+++ b/tests/reg_issue2478/reg_issue2478.fz
@@ -22,4 +22,23 @@
 # -----------------------------------------------------------------------
 
 x is
-b array x := []
+b1 array x := []
+b2 array x := [x,x]
+b3 := [x,x]
+c1 array unit := []
+c2 array unit := [unit,unit]
+c3 := [unit,unit]
+d1 array void   := []    # not sure if type propagation works here
+d2 array choice := []    # not sure if type propagation works here
+d3 := array void   0 (_ -> panic "*** not reached ***")
+d4 := array choice 0 (_ -> panic "*** not reached ***")
+say b1
+say b2
+say b3
+say c1
+say c2
+say c3
+say d1
+say d2
+say d3
+say d4

